TestProjects/VisualEffectGraph_HDRP/Assets/AllTests/Editor/Tests/VFXGUITests.cs

```diff
@@ -135,133 +97,230 @@ void CreateDataEdges(VFXContextController updateContext, List<VFXParameter> para
+            VFXContextType.Update,
+            VFXContextType.Output
+        }).ToArray();
 
         [Test]
-        public void CreateAllSpawnerBlocksTest()
+        public void CreateAllBlocksTest([ValueSource(nameof(kCreateAllBlockParam))] CreateAllBlockParam param)
         {
-            EditTestAsset(0);
-            CreateAllBlocks(VFXContextType.Spawner);
+            var viewController = StartEditTestAsset();
+            CreateAllBlocks(viewController, param.destContext, param.blocks);
         }
 
+        //Reduced selection for CreateAllDataEdgesTest, only testing block outputs
+        static CreateAllBlockParam[] kCreateAllBlockParamOutput = kCreateAllBlockParam.Where(o => o.destContext.model.contextType == VFXContextType.Output).ToArray();
         [Test]
-        public void CreateAllEventBlocksTest()
+        public void CreateAllDataEdgesTest([ValueSource(nameof(kCreateAllBlockParamOutput))] CreateAllBlockParam param)
         {
-            EditTestAsset(0);
-            CreateAllBlocks(VFXContextType.Event);
+            var viewController = StartEditTestAsset();
+            var contextController = CreateAllBlocks(viewController, param.destContext, param.blocks);
+            CreateDataEdges(viewController, contextController, CreateAllParameters(viewController));
         }
 
-        VFXContextController CreateAllBlocks(VFXContextType type)
+        VFXContextController CreateAllBlocks(VFXViewController viewController, VFXModelDescriptor<VFXContext> context, IEnumerable<VFXModelDescriptor<VFXBlock>> blocks)
         {
-            var initContextDesc = VFXLibrary.GetContexts().Where(t => t.model.contextType == type).First();
-
-            var newContext = m_ViewController.AddVFXContext(new Vector2(300, 2000), initContextDesc);
-
-            m_ViewController.ApplyChanges();
-
-            var contextController = m_ViewController.nodes.Where(t => t is VFXContextController && (t as VFXContextController).model == newContext).First() as VFXContextController;
+            var newContext = viewController.AddVFXContext(new Vector2(300, 2000), context);
+            viewController.ApplyChanges();
 
+            var contextController = viewController.nodes.Where(t => t is VFXContextController && (t as VFXContextController).model == newContext).First() as VFXContextController;
             Assert.AreEqual(contextController.model, newContext);
 
             // Adding every block compatible with an init context
-
             var newBlocks  = new List<VFXBlock>();
-            foreach (var block in VFXLibrary.GetBlocks().Where(t => t.AcceptParent(newContext)))
+            foreach (var block in blocks)
             {
                 var newBlock = block.CreateInstance();
                 contextController.AddBlock(0, newBlock);
                 newBlocks.Add(newBlock);
             }
 
-            m_ViewController.ApplyChanges();
-
+            viewController.ApplyChanges();
             foreach (var newBlock in newBlocks)
```

Comment:
It could probably be much faster (especially with 1k blocks) to do it like this (or you get the idea):
```csharp
var intersection = contextController.blockControllers.Select(x => x.model).Intersect(newBlocks).ToArray();
Assert.AreEqual(newBlocks.Count, interesction.Length);
Assert.AreEqual(newBlocks.Count, contextController.blockControllers.Count());
```

Reply:
This isn't exactly equivalen, it doesn't test a potential inversion of two model between two controller but it's really improbable. Thanks for this suggestion.

Reply:
Fixed at 7c1a938449fd6d60112bee5c2a9356de324c09ae